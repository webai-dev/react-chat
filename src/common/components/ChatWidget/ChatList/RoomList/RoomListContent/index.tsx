import React, {
  useState,
  useRef,
  useCallback,
  useContext,
  useEffect,
} from "react";
import Box from "@material-ui/core/Box";
import { makeStyles } from "@material-ui/styles";
import CircularProgress from "@material-ui/core/CircularProgress";
import ConnectionListItem from "../ConnectionListItem";
import { ConnectionsContext } from "../../../../../hooks/useConnections";
import { ConnectionItem } from "../../../../../../types/types";

export const useStyles = makeStyles({
  root: {
    flex: 1,
    overflow: "auto",
  },
  loaderContainer: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
});

interface RoomListContentProps {
  modalItem?: boolean;
  selectedMembers?: number[];
  setSelectedMembers?: React.Dispatch<React.SetStateAction<number[]>>;
  connectionStatus?: number;
}

/**
 * RoomList component
 * @param {{ RoomListContentProps }} props Properties
 * @description RoomList component consists of a list of RoomListItem components
 */
const RoomsListContent: React.FC<RoomListContentProps> = ({
  modalItem = false,
  selectedMembers = [],
  setSelectedMembers = () => undefined,
  connectionStatus = 1,
}) => {
  const classes = useStyles();

  const {
    connectionList,
    hasMore,
    loadNextPage,
    loadingList,
    loadingNextPage,
  } = useContext(ConnectionsContext);
  const containerRef = useRef() as React.MutableRefObject<HTMLInputElement>;
  const [modalRoomList, setModalRoomList] = useState<Array<ConnectionItem>>();

  useEffect(() => {
    let modalRooms: Array<ConnectionItem> = [];
    connectionList.forEach((room) => {
      if (connectionList) {
        modalRooms.push(room);
      }
    });
    setModalRoomList(modalRooms);
  }, [connectionList]);
  // Infinite scrolling code
  const observer = useRef<IntersectionObserver>();
  const lastConnectionElementRef = useCallback(
    (node) => {
      if (loadingNextPage) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting && hasMore) {
          loadNextPage();
          containerRef.current.scrollTop = containerRef.current.scrollHeight;
        }
      });
      if (node) observer.current.observe(node);
    },
    [loadingNextPage, hasMore, loadNextPage]
  );

  return (
    <div
      data-testid="room-list-content"
      className={classes.root}
      ref={containerRef}
    >
      {loadingList ? (
        <Box
          display="flex"
          height={410}
          justifyContent="center"
          alignItems="center"
        >
          <CircularProgress />
        </Box>
      ) : (
        <span>
          {modalRoomList &&
            modalRoomList.map((connection, index) => (
              <div
                key={index}
                ref={
                  connectionList.length === index + 1
                    ? lastConnectionElementRef
                    : null
                }
              >
                <ConnectionListItem
                  connectionStatus={connectionStatus}
                  connection={connection}
                  modalItem={modalItem}
                  selectedMembers={selectedMembers}
                  setSelectedMembers={setSelectedMembers}
                />
              </div>
            ))}

          {loadingNextPage && (
            <Box
              display="flex"
              flex={1}
              height={90}
              justifyContent="center"
              alignItems="center"
            >
              <CircularProgress />
            </Box>
          )}
        </span>
      )}
    </div>
  );
};

export default RoomsListContent;
