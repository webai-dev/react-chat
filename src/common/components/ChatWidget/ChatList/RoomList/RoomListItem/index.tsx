import React, { useContext } from "react";
import { makeStyles } from "@material-ui/styles";
import ChatImage from "../../../_Common/ChatImage";
import { RoomItem } from "../../../../../../types/types";

import Checkbox from "@material-ui/core/Checkbox";
import { CompanyContext } from "../../../../../hooks/useCompanies";
import { ChatRoomsContext } from "../../../../../hooks/useChatRooms";

// Global styles
export const useStyles = makeStyles({
  iconContainer: {
    paddingLeft: "15px",
    paddingRight: "12px",
    color: "#707070",
  },
  info: {
    marginLeft: 50,
  },
  name: {
    color: "#707070",
    fontSize: 16,
  },
  occupation: {
    color: "#707070",
    fontSize: 13,
  },
  root: {
    position: "relative",
    display: "flex",
    alignItems: "center",
    padding: "16px 22px 13px",
    cursor: "pointer",

    backgroundColor: (selected) => (selected ? "#4F8DCB1A" : "white"),
  },
  status: {
    position: "absolute",
    width: 12,
    height: 12,
    bodorRadius: 8,
    boxShadow: "0px 1px 3px #00000029",
    border: "1px solid #707070",
    borderRadius: 6,
    bottom: 0,
    right: 0,
  },
  unread: {
    width: 3,
    height: "100%",
    position: "absolute",
    backgroundColor: "#3870A0",
    left: 0,
    top: 0,
  },
});

interface RoomListItemProps {
  onClick?: () => void;
  team?: boolean;
  modal?: boolean;
  room: RoomItem;
  selectedRoomId?: number;
  modalItem?: boolean;
  selectedMembers?: number[];
  setSelectedMembers?: React.Dispatch<React.SetStateAction<number[]>>;
  connectionStatus: number;
}

/**
 * RoomListItem component
 * @param {RoomListItemProps} props Properties
 * @description RoomListItem component consists of data for one user/group in Chat Room
 */
const RoomListItem: React.FC<RoomListItemProps> = ({
  onClick = () => undefined,
  selectedRoomId = -1,
  room = {
    id: 0,
    lastActiveUsers: [],
    lastReadMessageID: 0,
    name: "",
    roomType: 0,
    unreadCount: 0,
  },
  modalItem = false,
  selectedMembers = [],
  setSelectedMembers = () => undefined,
  connectionStatus = 1,
}) => {
  const { id, name, lastMessage, roomType, unreadCount, lastActiveUsers } =
    room;
  const classes = useStyles(selectedRoomId === id);
  const { addChatRoom } = useContext(ChatRoomsContext);
  const { selectedCompany } = useContext(CompanyContext);

  let roomTitle = "";
  let roomSubTitle = "";

  if (roomType === 0) {
    roomTitle = name || "Group chat room";
    roomSubTitle = lastMessage ? lastMessage.message : "";
  } else {
    roomTitle = lastActiveUsers
      ? lastActiveUsers[0]?.firstName + " " + lastActiveUsers[0]?.lastName
      : "";
    roomSubTitle = lastActiveUsers
      ? lastActiveUsers[0]?.companyMembership
          .map((membership) => membership.jobTitle)
          .join(", ")
      : "";
  }

  const handleClickRoomListItem = () => {
    if (!modalItem) {
      addChatRoom(id, selectedCompany && selectedCompany.id);
      onClick();
    }
  };

  return (
    <div
      data-testid="room-list-item"
      className={classes.root}
      onClick={handleClickRoomListItem}
      style={
        modalItem
          ? {
              borderBottom: "1px solid rgba(0, 0, 0, 0.1)",
              marginRight: "80px",
            }
          : {}
      }
    >
      {unreadCount > 0 && <div className={classes.unread} />}
      {modalItem && (
        <Checkbox
          color="default"
          checked={
            selectedMembers.filter(
              (id) =>
                room.lastActiveUsers &&
                room.lastActiveUsers[0].id &&
                id === room.lastActiveUsers[0].id
            ).length > 0
          }
          onChange={(e) => {
            if (e.target.checked) {
              setSelectedMembers([
                ...selectedMembers,
                room.lastActiveUsers[0].id,
              ]);
            } else {
              let updatedMembers = selectedMembers.filter(
                (id) => id !== room.lastActiveUsers[0].id
              );
              setSelectedMembers(updatedMembers);
            }
          }}
          inputProps={{ "aria-label": "checkbox with default color" }}
        />
      )}

      <ChatImage connectionStatus={connectionStatus} room={room} />
      <div className={classes.info}>
        <div className={classes.name}>{roomTitle}</div>
        <div className={classes.occupation}>{roomSubTitle}</div>
      </div>
    </div>
  );
};

export default RoomListItem;
