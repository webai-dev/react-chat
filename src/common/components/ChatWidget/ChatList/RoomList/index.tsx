import React, {
  useState,
  useRef,
  useCallback,
  useContext,
  useEffect,
} from "react";
import Box from "@material-ui/core/Box";
import { makeStyles } from "@material-ui/styles";
import CircularProgress from "@material-ui/core/CircularProgress";
import RoomListItem from "./RoomListItem";
import { RoomListsContext } from "../../../../hooks/useRoomLists";
import { CompanyContext } from "../../../../hooks/useCompanies";
import { getRooms } from "../../../../services/rooms";
import { RoomListResponse } from "../../../../../types/types";
export const useStyles = makeStyles({
  root: {
    flex: 1,
    overflow: "auto",
  },
  loaderContainer: {
    display: "flex",
    justifyContent: "center",
    alignItems: "center",
  },
});

interface RoomListProps {
  connectionStatus: number;
}

/**
 * RoomList component
 * @param {{ RoomListProps }} props Properties
 * @description RoomList component consists of a list of RoomListItem components
 */
const RoomsList: React.FC<RoomListProps> = ({ connectionStatus = 1 }) => {
  const classes = useStyles();
  const [selectedRoomId, setSelectedRoomId] = useState(-1);
  const { selectedCompany } = useContext(CompanyContext);
  const {
    activeRoomList,
    loadNextPage,
    loadingList,
    loadingNextPage,
    companyRoomLists,
    setLoadingList,
  } = useContext(RoomListsContext);
  const containerRef = useRef() as React.MutableRefObject<HTMLInputElement>;

  useEffect(() => {
    if (connectionStatus === 1 && selectedCompany) {
      if (!activeRoomList) {
        setLoadingList(true);
        getRooms(selectedCompany.id, 0).then(
          (data: RoomListResponse | null) => {
            if (data) {
              companyRoomLists.push({
                companyId: selectedCompany.id,
                rooms: data.data,
                page: data.page,
                size: data.size,
                searchTerm: null,
                hasMore: data.totalPages > data.page,
              });

              setLoadingList(false);
            }
          }
        );
      } else {
        let updatedRoomsList = companyRoomLists.find(
          (c) => c.companyId === selectedCompany.id
        );
        if (updatedRoomsList) {
          setLoadingList(true);
          getRooms(selectedCompany.id, 0).then(
            (data: RoomListResponse | null) => {
              if (data && updatedRoomsList) {
                updatedRoomsList.companyId = selectedCompany.id;
                updatedRoomsList.rooms = data.data;
                updatedRoomsList.page = data.page;
                updatedRoomsList.size = data.size;
                updatedRoomsList.searchTerm = null;
                updatedRoomsList.hasMore = data.totalPages > data.page;

                setLoadingList(false);
              }
            }
          );
        }
        companyRoomLists.map((company) => {
          return company.companyId === selectedCompany.id
            ? updatedRoomsList
            : company;
        });
      }

      // updateActiveRooms();
    }
    // eslint-disable-next-line
  }, [connectionStatus]);

  // Infinite scrolling code
  const observer = useRef<IntersectionObserver>();
  const lastRoomElementRef = useCallback(
    (node) => {
      if (loadingNextPage) return;
      if (observer.current) observer.current.disconnect();
      observer.current = new IntersectionObserver((entries) => {
        if (entries[0].isIntersecting && activeRoomList?.hasMore) {
          loadNextPage();
          containerRef.current.scrollTop = containerRef.current.scrollHeight;
        }
      });
      if (node) observer.current.observe(node);
    },
    [loadingNextPage, activeRoomList?.hasMore, loadNextPage]
  );

  return (
    <div data-testid="room-list" className={classes.root} ref={containerRef}>
      {loadingList ? (
        <Box
          display="flex"
          height={410}
          justifyContent="center"
          alignItems="center"
        >
          <CircularProgress />
        </Box>
      ) : (
        <span>
          {activeRoomList?.rooms &&
            activeRoomList?.rooms.map((room, index) => (
              <div
                key={index}
                ref={
                  activeRoomList?.rooms.length === index + 1
                    ? lastRoomElementRef
                    : null
                }
              >
                <RoomListItem
                  connectionStatus={connectionStatus}
                  room={room}
                  selectedRoomId={selectedRoomId}
                  onClick={() => setSelectedRoomId(room.id)}
                />
              </div>
            ))}

          {loadingNextPage && (
            <Box
              display="flex"
              flex={1}
              height={90}
              justifyContent="center"
              alignItems="center"
            >
              <CircularProgress />
            </Box>
          )}
        </span>
      )}
    </div>
  );
};

export default RoomsList;
