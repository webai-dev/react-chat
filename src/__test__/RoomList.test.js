import React from "react";
import { render } from "@testing-library/react";
import "@testing-library/jest-dom";
import RoomList from "../common/components/ChatWidget/ChatList/RoomList";
import renderer from "react-test-renderer";

it("renders without crashing", async () => {
  const { getByTestId } = render(<RoomList />);
  const roomList = getByTestId("room-list");
  expect(roomList).toBeTruthy();
});

it("matches snapshot", async () => {
  const tree = renderer.create(<RoomList />).toJSON();
  expect(tree).toMatchSnapshot();
});
