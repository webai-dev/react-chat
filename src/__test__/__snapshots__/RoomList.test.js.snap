// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  className="makeStyles-root-3"
  data-testid="room-list"
>
  <span />
</div>
`;
